dart/dynamics/InverseKinematics.cpp

```diff
@@ -0,0 +1,1202 @@
+  mLastConfig.resize(0);
+}
+
+//==============================================================================
+InverseKinematics::TaskSpaceRegion::TaskSpaceRegion(
+    InverseKinematics* _ik,
+    const Properties& _properties,
+    bool _computeFromCenter)
+  : ErrorMethod(_ik, "EulerAngleXYZ", _properties),
+    mComputeErrorFromCenter(_computeFromCenter)
+{
+  // Do nothing
+}
+
+//==============================================================================
+std::unique_ptr<InverseKinematics::ErrorMethod>
+InverseKinematics::TaskSpaceRegion::clone(InverseKinematics* _newIK) const
+{
+  return std::unique_ptr<ErrorMethod>(
+        new TaskSpaceRegion(_newIK, mProperties, mComputeErrorFromCenter));
+}
+
+//==============================================================================
+Eigen::Vector6d InverseKinematics::TaskSpaceRegion::computeError()
+{
+  // This is a slightly modified implementation of the Berenson et al Task Space
+  // Region method found in "Task Space Regions: A Framework for
+  // Pose-Constrained Manipulation Planning".
+  //
+  // The modification is that we handle the computation of translational error
+  // independently from the computation of rotational error, whereas the work of
+  // Berenson computes them simultaneously using homogeneous transform matrices.
+  // Handling them in terms of their homogeneous matrices results in less direct
+  // paths while performing gradient descent, because it produces a pseudo-error
+  // in translation whenever rotational error exists.
+  //
+  // Also, we offer the option of always computing the error from the center of
+  // the Task Space Region, which will often allow the end effector to enter the
+  // valid constraint manifold faster than computing it from the edge of the
+  // Task Space Region.
+
+  // Use the target's transform with respect to its reference frame
+  const Eigen::Isometry3d& targetTf =
+      mIK->getTarget()->getRelativeTransform();
+  // Use the actual transform with respect to the target's reference frame
+  const Eigen::Isometry3d& actualTf =
+      mIK->getNode()->getTransform(mIK->getTarget()->getParentFrame());
+
+  // ^ This scheme makes it so that the bounds are expressed in the reference
+  // frame of the target
+
+  Eigen::Vector3d p_error = actualTf.translation() - targetTf.translation();
+  if(mIK->hasOffset())
+    p_error += actualTf.linear()*mIK->getOffset();
+
+  Eigen::Matrix3d R_error = actualTf.linear() * targetTf.linear().transpose();
+
+  Eigen::Vector6d displacement;
+  displacement.head<3>() = math::matrixToEulerXYZ(R_error);
+  displacement.tail<3>() = p_error;
+
+  Eigen::Vector6d error;
+  const Eigen::Vector6d& min = mProperties.mBounds.first;
+  const Eigen::Vector6d& max = mProperties.mBounds.second;
+  double tolerance = mIK->getSolver()->getTolerance();
+  for(int i=0; i<6; ++i)
+  {
+    if( displacement[i] < min[i] )
+    {
+      if(mComputeErrorFromCenter)
+      {
+        if(!std::isinf(max[i]))
```

Comment:
How about `std::isfinite()`? It would be more readable, and also `!std::isinf(NAN)` returns true so below logic wouldn't work for the case.